packages/jsapi-components/src/spectrum/utils/usePickerProps.ts

```diff
@@ -0,0 +1,187 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import {
+  ItemKey,
+  NormalizedItem,
+  NormalizedSection,
+  usePickerItemScale,
+} from '@deephaven/components';
+import { TableUtils } from '@deephaven/jsapi-utils';
+import Log from '@deephaven/log';
+import { usePromiseFactory } from '@deephaven/react-hooks';
+import { PICKER_TOP_OFFSET } from '@deephaven/utils';
+import useFormatter from '../../useFormatter';
+import type { PickerWithTableProps } from '../PickerProps';
+import { getItemKeyColumn, getItemLabelColumn } from './itemUtils';
+import { useItemRowDeserializer } from './useItemRowDeserializer';
+import { useGetItemIndexByValue } from '../../useGetItemIndexByValue';
+import useSearchableViewportData from '../../useSearchableViewportData';
+
+const log = Log.module('jsapi-components.usePickerProps');
+
+/** Props that are derived by `usePickerProps`. */
+export type UsePickerDerivedProps = {
+  normalizedItems: (NormalizedItem | NormalizedSection)[];
+  showItemIcons: boolean;
+  getInitialScrollPosition: () => Promise<number | null>;
+  onChange: (key: ItemKey | null) => void;
+  onScroll: (event: Event) => void;
+  onSearchTextChange: (searchText: string) => void;
+};
+
+/** 
+ * Props that are passed through untouched. (should exclude all of the
+ * destructured props passed into `usePickerProps` that are not in the spread
+ * ...props)
+) */
+export type UsePickerPassthroughProps<TProps> = Omit<
+  PickerWithTableProps<TProps>,
+  | 'table'
+  | 'keyColumn'
+  | 'labelColumn'
+  | 'iconColumn'
+  | 'settings'
+  | 'onChange'
+  | 'onSelectionChange'
+>;
+
+/** Props returned by `usePickerProps` hook. */
+export type UsePickerProps<TProps> = UsePickerDerivedProps &
+  UsePickerPassthroughProps<TProps>;
+
+export function usePickerProps<TProps>({
+  table: tableSource,
+  keyColumn: keyColumnName,
+  labelColumn: labelColumnName,
+  iconColumn: iconColumnName,
+  settings,
+  onChange,
+  onSelectionChange,
+  ...props
+}: PickerWithTableProps<TProps>): UsePickerProps<TProps> {
+  const { itemHeight } = usePickerItemScale();
+
+  const { getFormattedString: formatValue, timeZone } = useFormatter(settings);
+
+  // `null` is a valid value for `selectedKey` in controlled mode, so we check
+  // for explicit `undefined` to identify uncontrolled mode.
+  const isUncontrolled = props.selectedKey === undefined;
+  const [uncontrolledSelectedKey, setUncontrolledSelectedKey] = useState<
+    ItemKey | null | undefined
+  >(props.defaultSelectedKey);
+
+  // Copy table so we can apply filters without affecting the original table.
+  // (Note that this call is not actually applying any filters. Filter will be
+  // applied in `useSearchableViewportData`.)
+  const { data: tableCopy } = usePromiseFactory(
```

Comment:
@mofojed for reference, I think this is where we could apply addition filters as part of @AkshatJawne partition table work

Reply:
I guess it might make sense to just add it to this PR, but I'm fine either way